docs/proposals/auth.md

```diff
@@ -0,0 +1,128 @@
+Heapster Authentication and Authorization
+=========================================
+
+Current State
+-------------
+
+Heapster authentication is based on checking client certificates against a CA.
+Authorization is either just a static list of names, or non-existant.  In
+either case, anyone authenticated is authorized to perform any operation.  With
+push metrics, the proposed plan was to allow adding a second CA, to allow
+separating authorization for model and Oldtimer queries from authorization to
+push metrics.  However, this is suboptimal, since it also separates
+authentication.
+
+Requirements
+------------
+
+Heapster should be able to separate authentication from authorization, and
+the authorization interface should be flexible enough to support authorization
+implementations that are able to make the following distinctions:
+
+- containing queries to a particular namespace: we should be able to only allow
+  pods to query for metrics within their own namespace.
+
+- separating querying from pushing: we should allow certain users to push, and
+  certain users to query, but not necessarily tie the two together
+
+- allowing different push scope permissions: we should be able to specify that
+  some users are only allowed to push metrics for the pods within their own
+  namespace (or given sufficient information in the future, possibly their own
+  RC), while others should be allowed to push for any object in the cluster.
+
+Proposed Solution
+-----------------
+
+Heapster will adopt an authentication and authorization model based around
+querying the Kubernetes model.  Authentication will support either client
+certificate and tokens using the Kubernetes API server tokenreview
+endpoint. Authorization will be done using the Kubernetes API server
+subjectaccessreviews endpoint.
+
+### Authentication ###
+
+Authentication will support both client certificates and tokens.  Client
+certificates will be supported by checking against a configurable CA (this
+will generally be the cluster CA that gets injected into the Heapster
+pod), while tokens will be checked against the Kubernetes API server using
+a `TokenReview`.
+
+The configuration will look as such:
+
+- `--authn-ca=/path/to/ca.crt`
+- `--auth-apiserver=https://$KUBE`
+
+### Authorization ###
+
+Authorization will work by matching Heapster concepts to Kubernetes RBAC
+concepts, and then performing `SubjectAccessReview` requests against the
+API server configured in `--auth-apiserver`.
+
+Concepts will map as such:
```

Comment:
That won't work when Heapster gains the ability to support metrics on objects other than pods.  How do you distinguish between querying for metrics on an object in two different API groups if you can't pay attention to the SubjectAccessReview's API group?